Implement the Apple Pay JS API for deferred payment requests

squelart opened the pull request:
#### 1fdc00a92679f0507c9a3a22f156f9491c62bdb4
<pre>
Implement the Apple Pay JS API for deferred payment requests
<a href="https://bugs.webkit.org/show_bug.cgi?id=253219">https://bugs.webkit.org/show_bug.cgi?id=253219</a>
rdar://104562407

Reviewed by Aditya Keerthi and Dean Jackson.

The new Apple Pay JS API ApplePayDeferredPaymentRequest corresponds to PKDeferredPaymentRequest <a href="https://developer.apple.com/documentation/passkit/pkdeferredpaymentrequest">https://developer.apple.com/documentation/passkit/pkdeferredpaymentrequest</a>

Payment requests now accept an optional property &apos;deferredPaymentRequest&apos;, which may contain an item with a deferred payment timing, and optionally a free cancellation date. This may be used for e.g.: pre-orders.

* LayoutTests/http/tests/paymentrequest/ApplePayModifier-deferredPaymentRequest.https-expected.txt: Added.
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-deferredPaymentRequest.https.html: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/applepay/ApplePayDeferredPaymentRequest.h: Copied from Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.idl.
* Source/WebCore/Modules/applepay/ApplePayDeferredPaymentRequest.idl: Copied from Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.idl.
* Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.h:
* Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.idl:
* Source/WebCore/Modules/applepay/ApplePayFeature.h:
* Source/WebCore/Modules/applepay/ApplePayFeature.idl:
* Source/WebCore/Modules/applepay/ApplePayPaymentRequest.h:
* Source/WebCore/Modules/applepay/ApplePayPaymentRequest.idl:
* Source/WebCore/Modules/applepay/ApplePaySession.cpp:
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
(WebCore::ApplePaySessionPaymentRequest::deferredPaymentRequest const):
(WebCore::ApplePaySessionPaymentRequest::setDeferredPaymentRequest):
* Source/WebCore/Modules/applepay/cocoa/PaymentAPIVersionCocoa.mm:
(WebCore::PaymentAPIVersion::current):
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayModifier.h:
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayModifier.idl:
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
(WebCore::ApplePayPaymentHandler::show):
(WebCore::validate):
(WebCore::ApplePayPaymentHandler::shippingAddressUpdated):
(WebCore::ApplePayPaymentHandler::shippingOptionUpdated):
(WebCore::ApplePayPaymentHandler::paymentMethodUpdated):
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h:
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/testing/MockPaymentCoordinator.cpp:
(WebCore::MockPaymentCoordinator::showPaymentUI):
(WebCore::MockPaymentCoordinator::completeShippingMethodSelection):
(WebCore::MockPaymentCoordinator::completeShippingContactSelection):
(WebCore::MockPaymentCoordinator::completePaymentMethodSelection):
(WebCore::MockPaymentCoordinator::completeCouponCodeChange):
* Source/WebCore/testing/MockPaymentCoordinator.h:
* Source/WebCore/testing/MockPaymentCoordinator.idl:
* Source/WebKit/Platform/cocoa/PaymentAuthorizationPresenter.mm:
(WebKit::PaymentAuthorizationPresenter::completePaymentMethodSelection):
(WebKit::PaymentAuthorizationPresenter::completeShippingContactSelection):
(WebKit::PaymentAuthorizationPresenter::completeShippingMethodSelection):
(WebKit::PaymentAuthorizationPresenter::completeCouponCodeChange):
* Source/WebKit/Shared/ApplePay/DeferredPaymentRequest.h: Copied from Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.idl.
* Source/WebKit/Shared/ApplePay/cocoa/DeferredPaymentRequestCocoa.mm: Added.
(WebKit::platformDeferredPaymentRequest):
* Source/WebKit/Shared/ApplePay/cocoa/PaymentTokenContextCocoa.mm:
(WebKit::toDecimalNumber): Deleted.
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::WebPaymentCoordinatorProxy::platformPaymentRequest):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder&lt;WebCore::ApplePaySessionPaymentRequest&gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::ApplePaySessionPaymentRequest&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/261216@main">https://commits.webkit.org/261216@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/293f21a0eab0b200c7debadba37ecb66eb90f903

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110796 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19882 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2332 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119615 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114761 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21283 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10987 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/2020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116546 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15858 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99021 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103207 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97818 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30687 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44256 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/99444 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12474 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32023 "Found 1 new test failure: webanimations/animation-of-accelerated-property-after-non-accelerated-property.html (failure)") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/10584 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13043 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8999 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/100528 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18426 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31354 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7786 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14997 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/108558 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26761 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/Modules/applepay/cocoa/PaymentAPIVersionCocoa.mm:
```diff
@@ -34,8 +34,11 @@
+    static const unsigned current = [] {
+#if ENABLE(APPLE_PAY_DEFERRED_PAYMENTS)
+        // FIXME: No public version 15 yet, so should Deferred Payments be in 15 like Features?
+        return 16;
```
The version number should be frozen.

See https://github.com/WebKit/WebKit/pull/2793 for more details.

squelart replied:
Oh I see, excellent!
Should I add a comment there, so that others won't be tempted to change it again?

squelart replied:
I've added a comment, but let me know if you'd prefer I didn't touch this file.

pxlcoder replied:
A comment is a good idea!

I would say: 

> This version number should not be changed anymore, features can now be found in method.data.features.

Rather than:

> This version number should not need to change anymore, features may now be found in method.data.features.

pxlcoder commented on Source/WebCore/Modules/applepay/ApplePayDeferredPaymentRequest.h:
```diff
@@ -0,0 +1,49 @@
+#if ENABLE(APPLE_PAY_DEFERRED_PAYMENTS)
+
+#include <WebCore/ApplePayLineItem.h>
+#include <optional>
```
This include seems unnecessary?

pxlcoder commented on Source/WebCore/Modules/applepay/ApplePayDeferredPaymentRequest.h:
```diff
@@ -0,0 +1,49 @@
+struct ApplePayDeferredPaymentRequest final {
+    String billingAgreement;
+    ApplePayLineItem deferredBilling; // required, must be `paymentTiming: "deferred"
+    WallTime freeCancellationDate { WallTime::nan() };
```
It might be nice to use `Markable` here, but I see that the other Apple Pay code is written this way.

squelart replied:
I'll keep it consistent for now, but I've filed rdar://106132745 to look into it later.

pxlcoder commented on Source/WebCore/Modules/applepay/ApplePaySession.cpp:
```diff
@@ -64,6 +65,7 @@
 #include "SecurityOrigin.h"
 #include "Settings.h"
 #include "UserGestureIndicator.h"
+#include <CoreFoundation/CFTimeZone.h>
```
It seems wrong to include a `CoreFoundation` header at this layer of the code, even though `ENABLE(APPLE_PAY)` is only true on Cocoa platforms.

It might be better to add a helper in WTF, using icu. Something like `canonicalizeTimeZoneName` could be moved from JSC into WTF.

Thoughts @hortont424?

squelart replied:
Unless it's a trivial change, I'd appreciate if we could defer this change to a later test phase (I'd file a radar and add a FIXME). Please let me know.

pxlcoder replied:
I think it would just be moving (re-using) the logic in `canonicalizeTimeZoneName`. My understanding is that if that method returns a `null` String, the time zone is invalid.

Maybe @Constellation has thoughts about where the helper could live?

squelart replied:
In this latest update, I've just copied the `canonicalizeTimeZoneName` code where I needed it, with a FIXME to later move it to WTF, because I think it would probably need some careful consideration and testing. (Happy to open a radar if you agree.)
But please let me know if you'd prefer another location or solution right now.

squelart replied:
New update, now using `WTF::isTimeZoneValid`, thank you for the hint.

pxlcoder commented on Source/WebKit/Shared/ApplePay/cocoa/PaymentTokenContextCocoa.mm:
```diff
@@ -38,13 +39,6 @@
 namespace WebKit {
 using namespace WebCore;
 
-static NSDecimalNumber *toDecimalNumber(const String& amount)
-{
-    if (!amount)
-        return NSDecimalNumber.zero;
-    return [NSDecimalNumber decimalNumberWithString:amount locale:@{ NSLocaleDecimalSeparator : @"." }];
-}
-
```
Is this change necessary in this patch? I see a couple of other places that have this code duplicated, so maybe it's better we just address it in a follow-up.

squelart replied:
It was necessary here, because I'm adding files, which changes how files are unified, causing this one to start clashing with another one! It can't be a follow-up, but it could be in a prior one, if you really wanted this change out of this patch.

pxlcoder replied:
I see! This is fine for now. We can always do a follow-up afterwards.

pxlcoder commented on Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
```diff
@@ -546,6 +550,35 @@ static ExceptionOr<void> validate(const ApplePayModifier& applePayModifier)
+
+        if (auto& tokenNotificationURL = deferredPaymentRequest->tokenNotificationURL; !tokenNotificationURL.isNull() && !URL { tokenNotificationURL }.isValid())
+            return Exception(TypeError, makeString('"', tokenNotificationURL, "\" is not a valid URL."));
+    }
```
There seems to be a lot of duplication between this and `convertAndValidate`. Can we share the code?

grorg replied:
Agreed.

squelart replied:
Ok, I've managed to combine them.
I tried to put them in a new file "ApplePayDeferredPaymentRequest.cpp", but the unified build got confused -- please don't make me fix these things today! Happy to file a follow-up radar, which should also try to de-duplicate the validation of other structs.

pxlcoder replied:
Okay, we can do that in a follow-up.

squelart replied:
I've filed rdar://106207791 (De-duplicate payment validation code)

pxlcoder commented on Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
```diff
@@ -546,6 +550,35 @@ static ExceptionOr<void> validate(const ApplePayModifier& applePayModifier)
+                return Exception(TypeError, "Unexpected 'freeCancellationDateTimeZone' when 'freeCancellationDate' is not set."_s);
+        } else if (auto& freeCancellationDateTimeZone = deferredPaymentRequest->freeCancellationDateTimeZone; freeCancellationDateTimeZone.isEmpty())
+            return Exception(TypeError, "Missing 'freeCancellationDateTimeZone' when 'freeCancellationDate' is set."_s);
+        else if (!CFTimeZoneCreateWithName(nullptr, freeCancellationDateTimeZone.createCFString().get(), false))
```
This should use `isTimeZoneValid ` rather than `CFTimeZoneCreateWithName`. Hopefully addressed by my other comment.

grorg replied:
Yeah, and generally we try to avoid Apple-specific stuff like CF outside of platform. (Not that important here considering it is an Apple-specific feature, and a module)

pxlcoder commented on Source/WebKit/Shared/ApplePay/cocoa/PaymentTokenContextCocoa.mm:
```diff
@@ -28,6 +28,7 @@
 
 #if HAVE(PASSKIT_MULTI_MERCHANT_PAYMENTS)
 
+#import "WebPaymentCoordinatorProxyCocoa.h" // for toDecimalNumber, see FIXME in that header
```
We don't normally leave comments next to imports, I would remove this.

grorg commented on Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
```diff
@@ -195,16 +201,20 @@ class ApplePaySessionPaymentRequest {
 
 #if ENABLE(APPLE_PAY_RECURRING_PAYMENTS)
-    std::optional<ApplePayAutomaticReloadPaymentRequest> m_automaticReloadPaymentRequest;
+    std::optional<ApplePayRecurringPaymentRequest> m_recurringPaymentRequest;
```
Funny that these were the wrong way around. But it also makes me wonder if we need the ENABLE guards any more.

squelart replied:
Luckily those two features were added at the same time, so it just worked™. 😄

grorg commented on Source/WebKit/Shared/ApplePay/cocoa/DeferredPaymentRequestCocoa.mm:
```diff
@@ -0,0 +1,59 @@
+
+RetainPtr<PKDeferredPaymentRequest> platformDeferredPaymentRequest(const ApplePayDeferredPaymentRequest& webDeferredPaymentRequest)
+{
+    auto pkDeferredPaymentRequest = adoptNS([PAL::allocPKDeferredPaymentRequestInstance() initWithPaymentDescription:webDeferredPaymentRequest.paymentDescription deferredBilling:platformDeferredSummaryItem(webDeferredPaymentRequest.deferredBilling) managementURL:[NSURL URLWithString:webDeferredPaymentRequest.managementURL]]);
```
Small nit: people tend to split these calls over multiple lines for readability.

grorg reviewed (approved):
r+ assuming the comments are addressed.

pxlcoder commented on Source/WebCore/Modules/applepay/ApplePaySession.cpp:
```diff
@@ -67,6 +67,10 @@
 #include <wtf/RunLoop.h>
 
+#if ENABLE(APPLE_PAY_DEFERRED_PAYMENTS)
+# include <wtf/DateMath.h>
```
Nit: Remove extra space before `include`.

pxlcoder commented on Source/WebCore/Modules/applepay/ApplePaySession.cpp:
```diff
@@ -67,6 +67,10 @@
 
+#if ENABLE(APPLE_PAY_DEFERRED_PAYMENTS)
+# include <wtf/DateMath.h>
+#endif // ENABLE(APPLE_PAY_DEFERRED_PAYMENTS)
```
Nit: Trailing comment is not really necessary here, since it's a very short section.

squelart committed:
Implement the Apple Pay JS API for deferred payment requests
https://bugs.webkit.org/show_bug.cgi?id=253219
rdar://104562407

Reviewed by Aditya Keerthi and Dean Jackson.

The new Apple Pay JS API ApplePayDeferredPaymentRequest corresponds to PKDeferredPaymentRequest https://developer.apple.com/documentation/passkit/pkdeferredpaymentrequest

Payment requests now accept an optional property 'deferredPaymentRequest', which may contain an item with a deferred payment timing, and optionally a free cancellation date. This may be used for e.g.: pre-orders.

* LayoutTests/http/tests/paymentrequest/ApplePayModifier-deferredPaymentRequest.https-expected.txt: Added.
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-deferredPaymentRequest.https.html: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/applepay/ApplePayDeferredPaymentRequest.h: Copied from Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.idl.
* Source/WebCore/Modules/applepay/ApplePayDeferredPaymentRequest.idl: Copied from Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.idl.
* Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.h:
* Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.idl:
* Source/WebCore/Modules/applepay/ApplePayFeature.h:
* Source/WebCore/Modules/applepay/ApplePayFeature.idl:
* Source/WebCore/Modules/applepay/ApplePayPaymentRequest.h:
* Source/WebCore/Modules/applepay/ApplePayPaymentRequest.idl:
* Source/WebCore/Modules/applepay/ApplePaySession.cpp:
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
(WebCore::ApplePaySessionPaymentRequest::deferredPaymentRequest const):
(WebCore::ApplePaySessionPaymentRequest::setDeferredPaymentRequest):
* Source/WebCore/Modules/applepay/cocoa/PaymentAPIVersionCocoa.mm:
(WebCore::PaymentAPIVersion::current):
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayModifier.h:
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayModifier.idl:
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
(WebCore::ApplePayPaymentHandler::show):
(WebCore::validate):
(WebCore::ApplePayPaymentHandler::shippingAddressUpdated):
(WebCore::ApplePayPaymentHandler::shippingOptionUpdated):
(WebCore::ApplePayPaymentHandler::paymentMethodUpdated):
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h:
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/testing/MockPaymentCoordinator.cpp:
(WebCore::MockPaymentCoordinator::showPaymentUI):
(WebCore::MockPaymentCoordinator::completeShippingMethodSelection):
(WebCore::MockPaymentCoordinator::completeShippingContactSelection):
(WebCore::MockPaymentCoordinator::completePaymentMethodSelection):
(WebCore::MockPaymentCoordinator::completeCouponCodeChange):
* Source/WebCore/testing/MockPaymentCoordinator.h:
* Source/WebCore/testing/MockPaymentCoordinator.idl:
* Source/WebKit/Platform/cocoa/PaymentAuthorizationPresenter.mm:
(WebKit::PaymentAuthorizationPresenter::completePaymentMethodSelection):
(WebKit::PaymentAuthorizationPresenter::completeShippingContactSelection):
(WebKit::PaymentAuthorizationPresenter::completeShippingMethodSelection):
(WebKit::PaymentAuthorizationPresenter::completeCouponCodeChange):
* Source/WebKit/Shared/ApplePay/DeferredPaymentRequest.h: Copied from Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.idl.
* Source/WebKit/Shared/ApplePay/cocoa/DeferredPaymentRequestCocoa.mm: Added.
(WebKit::platformDeferredPaymentRequest):
* Source/WebKit/Shared/ApplePay/cocoa/PaymentTokenContextCocoa.mm:
(WebKit::toDecimalNumber): Deleted.
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::WebPaymentCoordinatorProxy::platformPaymentRequest):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder<WebCore::ApplePaySessionPaymentRequest>::encode):
(IPC::ArgumentCoder<WebCore::ApplePaySessionPaymentRequest>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/261216@main
